### PR TITLE
sqm-scripts: Fixes minor errors

### PR DESCRIPTION
**Pull request by rfrht**
Hi Toke,

As promised, here is the new commit set. These fixes some minor errors or issues that I have stumbled when working on my debian box.

Hope that helps.

### CODE REVIEW

**Comment by moeller0** [comment]
Hi Rodrigo,

> On Mar 26, 2016, at 14:19 , Rodrigo Freire <notifications@github.com> wrote:
> 
> Hi Toke,
> 
> As promised, here is the new commit set. These fixes some minor errors or issues that I have stumbled when working on my debian box.
> 
> Hope that helps.
> 
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/tohojo/sqm-scripts/pull/34
> 
> Commit Summary
> 
> 	• Example: Describe the up/downlink units

	This has no side effects and better documents the variables, I am all for it.

> 	• sqm-bin: Print a error if no ifs are defined

	This I believe will not run on openwrt, so this might have real issues (that I can not test myself)

> 	• stop-sqm: Source the interface before stop

	This supposed error does not show up on openwrt, so this might require a bit more research (your fix might be correct, but the diagnosis certainly is not sufficiently thorough for my taste).

> 	• Auxiliary scripts: Fix shell lingo

	Do you have a link regarding strict shell requirements? I assume there might be more of those hidden in functions.sh

> 	• run-openwrt: Fix lib directory

	I believe this one is wrong, as far as I can see we need to source openwrt’s /lib/functions.sh top have working config_foreach() and friends available. I believe you have not tested that change with a real openwrt. I have not yet looked at your other changes… I will try to find time to look through them later… But could you try to check whether your observed failure cases are specific to your distribution? Not that it is my decision, but I am uneasy with fixes that have had no testing under openwrt (which I assume the main platform for sqm-scripts right now). And if openwrt does work but $OTHER_DISTRO does not it might mean we need to fix things, and your fix finally might be the correct thing to do (excluding the run-openwrt change here)…

Best Regards
	Sebastian

**Comment by rfrht** [comment]
Hi Mœller,

> > Example: Describe the up/downlink units

>This has no side effects and better documents the variables, I am all for it.

Thanks!

> > sqm-bin: Print a error if no ifs are defined

> This I believe will not run on openwrt, so this might have real issues (that I can not test myself)

I don't envision why would that not run in `openwrt`. It's just a `else` condition and a bunch of echoes.

> > stop-sqm: Source the interface before stop

> This supposed error does not show up on openwrt, so this might require a bit more research (your fix might be correct, but the diagnosis certainly is not sufficiently thorough for my taste).

Sure. Let me show the root casuse. Check [this pastebin](http://pastebin.com/b5L7WxNi). See line 162:

~~~
+ [ ! -f /usr/lib/sqm/ ]
~~~

This is expanded from the following code snippet in `src/stop-sqm`:

~~~
if [ ! -f "${SQM_LIB_DIR}/$SCRIPT" ]; then
    sqm_error "SQM script ${SCRIPT} not found!"
    exit 1
fi
~~~

The script has no reference of where does the `$SCRIPT` variable comes from. Thus, sourcing the running interface value and populating accordingly the variable.

> > Auxiliary scripts: Fix shell lingo

> Do you have a link regarding strict shell requirements? I assume there might be more of those hidden in functions.sh

In shell parsers, only bash-variant parses `==` as a `=`. Check [this thread](http://unix.stackexchange.com/questions/16109/bash-double-equals-vs-eq). 
This change, besides fixing a bug, will make the code more universal and *correctly portable*. To assess the feasibility, do the following test. It will work on `bash`-derived shells. Will fail on `dash`.

~~~
[root@rasp-rodhome ~]# sh -s
# if [ 1 == 1 ] ; then echo blah ; fi
sh: 1: [: 1: unexpected operator
# if [ "1" == "1" ] ; then echo blah ; fi
sh: 2: [: 1: unexpected operator
# if [ 1 = 1 ] ; then echo blah ; fi
blah
# 
~~~

> > run-openwrt: Fix lib directory

> I believe this one is wrong [...]

Ok, I'll happily let it go, can't test, since I'm no `openWRT` user. But the `pwd` caught my attention.

In a nutshell, besides this last patch, I think that all four proposed patches are compatible with your embedded system.

Please reconsider.

With my warmest regard,

- Rodrigo.

**Comment by moeller0** [comment]
Hello Rodrigo,

> On Mar 26, 2016, at 16:40 , Rodrigo Freire <notifications@github.com> wrote:
> 
> Hi Mœller,
> 
> Example: Describe the up/downlink units
> 
> This has no side effects and better documents the variables, I am all for it.
> 
> Thanks!
> 
> sqm-bin: Print a error if no ifs are defined
> 
> This I believe will not run on openwrt, so this might have real issues (that I can not test myself)
> 
> I don't envision why would that not run in openwrt. It’s just a else condition and a bunch of echoes.

	sqm-bin is not part of sqm-scripts for openwrt, only for full linux distributions, that is all I wanted to convey, I can not test your change and hence have nothing to say.

> 
> stop-sqm: Source the interface before stop
> 
> This supposed error does not show up on openwrt, so this might require a bit more research (your fix might be correct, but the diagnosis certainly is not sufficiently thorough for my taste).
> 
> Sure. Let me show the root casuse. Check this pastebin. See line 162:
> 
> + [ ! -f /usr/lib/sqm/ ]
> 
> This is expanded from the following code snippet in src/stop-sqm:
> 
> if [ ! -f "${SQM_LIB_DIR}/$SCRIPT" ]; then
>     sqm_error "SQM script ${SCRIPT} not found!"
>     exit 1
> fi
> 
> The script has no reference of where does the $SCRIPT variable comes from. Thus, sourcing the running interface value and populating accordingly the variable.

	In openwrt sqm-stop is called from run-openwrt.sh, which will “export $SCRIPT” and call stop-sqm. Obviously then $IFACE is defined already and yes run-openwrt.sh sources an interface’s state file to get access to $SCRIPT, so maybe generic Linux really needs changes here, but openwrt should work out of the box already.

> 
> Auxiliary scripts: Fix shell lingo
> 
> Do you have a link regarding strict shell requirements? I assume there might be more of those hidden in functions.sh
> 
> In shell parsers, only bash-variant parses == as a =. Check this thread. 
> This change, besides fixing a bug, will make the code more universal and correctly portable. To assess the feasibility, do the following test. It will work on bash-derived shells. Will fail on dash.

	Personally I would love to require a decent shell and be done with ;) Just kidding, since sqm-scripts does very little there is no need to not run well on posix shells.

> 
> [root@rasp-rodhome ~]# sh -s
> # if [ 1 == 1 ] ; then echo blah ; fi
> sh: 1: [: 1: unexpected operator
> # if [ "1" == "1" ] ; then echo blah ; fi
> sh: 2: [: 1: unexpected operator
> # if [ 1 = 1 ] ; then echo blah ; fi
> blah
> # 

	Hewre is from openwrt:
root@nacktmulle:~# if [ 1 == 1 ] ; then echo blah ; fi
blah
root@nacktmulle:~# if [ "1" == "1" ] ; then echo blah ; fi
blah
root@nacktmulle:~# if [ 1 = 1 ] ; then echo blah ; fi
blah
root@nacktmulle:~# 

so this explains why I have never seen this, openwrt uses the busybox shell, which as it seems is sort of bash-lite. It also means that your change should work well for openwrt.



> run-openwrt: Fix lib directory
> 
> I believe this one is wrong [...]
> 
> Ok, I'll happily let it go, can't test, since I'm no openWRT user. But the pwd caught my attention.

	Well, not my call, but I believe changes that never received any testing should be marked as such, they might still be fine and acceptable, but it will make people look more carefully…


> 
> In a nutshell, besides this last patch, I think that all four proposed patches are compatible with your embedded system.
> 
> Please reconsider.

	Oh, Toke is the person to convince, I just try to be helpful from the sidelines.

> 
> With my warmest regard,
> 
> 	• Rodrigo.

Best Regards
	Sebastian

**Comment by tohojo** [comment]
Thanks! This all looks reasonable, with a few exceptions (see below).  If you could fix those, I'll be happy to merge this.

1. Please add Signed-off-by lines to each commit (passing -s to git commit will do this for you).
2. The shell script fix should probably be -eq rather than '=' since we're trying to do a numeric comparison.
3. For the stop-sqm logic I'd prefer it if you check that the file exists before sourcing it. And reuse the STATE_FILE variable that is already set. Adding a bailout if SCRIPT is not set with a nicer error message afterwards is then probably a good idea (in case the file doesn't exist and we don't get SCRIPT set from somewhere else).
4. The last commit is wrong; /lib/functions.sh is an openwrt-specific file, not part of sqm-scripts. Just get rid of that commit.


If you just change the commits and force push them to your own repository, you can keep this pull request open :)

Cheers,
-Toke
